text/0000-or-patterns.md

```diff
@@ -0,0 +1,858 @@
+```
+
+And for `while let` expressions:
+
+```rust
+expr_while_let : maybe_label WHILE LET top_pat '=' expr_nostruct block ;
+```
+
+For `for` loop expressions we now have:
+
+```rust
+expr_for : maybe_label FOR top_pat IN expr_nostruct block ;
+```
+
+For `match` expressions we now have:
+
+```rust
+expr_match : MATCH expr_nostruct '{' match_clause* nonblock_match_clause? '}' ;
+match_clause : nonblock_match_clause ',' | block_match_clause ','? ;
+nonblock_match_clause : match_arm (nonblock_expr | block_expr_dot) ;
+block_match_clause : match_arm (block | block_expr) ;
+
+match_arm : maybe_outer_attrs top_pat (IF expr_nostruct)? FAT_ARROW ;
+```
+
+In other words, in all of the contexts where a pattern is currently accepted,
+the compiler will now accept pattern alternations of form `p | q` where
+`p` and `q` are arbitrary patterns.
+
+For the patterns of `fn` arguments we now have:
+
+```rust
+param : pat<no_top_alt> ':' ty_sum ;
+```
+
+For closures we now have:
+
+```rust
+inferrable_param : pat<no_top_alt> maybe_ty_ascription ;
+```
+
+Finally, `pat` macro fragment specifiers will also match the `pat<no_top_alt>`
+production as opposed to `pat<allow_top_alt>`.
+
+### Error messages
+
+As previously noted, the precedence of the operator `|` is lower than that of
+the operator `@`. This results in `i @ p | q` being interpreted as `(i @ p) | q`.
+In turn, this would result in an error because `i` is not defined in all
+alternations. An example:
+
+```rust
+fn main() {
+    match 1 {
+        i @ 0 | 1 => {},
+    }
+}
+```
+
+This would result in:
+
+```rust
+error[E0408]: variable `i` is not bound in all patterns
+ --> src/main.rs:3:17
+  |
+3 |         i @ 0 | 1 => {},
+  |         -       ^ pattern doesn't bind `i`
+  |         |
+  |         variable not in all patterns
+```
+
+However, it is quite likely that a user who wrote `i @ p | q` wanted the
+semantics of `i @ (p | q)` because it would be the only thing that would
+be a well formed pattern. To guide the user on the way, we recommend special
+casing the error message for such circumstances with for example:
+
+```rust
+error[E0408]: variable `i` is not bound in all patterns
+ --> src/main.rs:3:17
+  |
+3 |         i @ 0 | 1 => {},
+  |         -       ^ pattern doesn't bind `i`
+  |         |
+  |         variable not in all patterns
+  |
+  | hint: if you wanted `i` to cover both cases, try adding parentheses around:
+  |
+  |         i @ 0 | 1
+  |             ^^^^^
+```
+
+The particular design of such an error message is left open to implementations.
+
+## Static semantics
+
+1. Given a pattern `p | q` at some depth for some arbitrary patterns `p` and `q`,
+   the pattern is considered ill-formed if:
+
+   + the type inferred for `p` does not unify with the type inferred for `q`, or
+   + the same set of bindings are not introduced in `p` and `q`, or
+   + the type of any two bindings with the same name in `p` and `q` do not unify
+     with respect to types or binding modes.
```

Comment:
Do coercions apply, or do they need to unify exactly?

Reply:
Summary from Discord conversation:

...So according to the coercion rules, if `match` is a coercion site, then the following should work according to the [type coercion rules](https://doc.rust-lang.org/reference/type-coercions.html):

```rust
fn main() {
    enum E<'a> {
        A(&'a u8),
        B(&'a mut u8),
    }

    match unimplemented!() {
        E::A(x) | E::B(x) => {
            let x: &u8 = x;
        }
    }
}
```

However; `match` is not coercion site, so possibly the coercions would apply for `let`.
For now, I think we can be conservative and require an exact match and relax it later if need be / it makes sense.

In relation to type ascription, one could reasonably expect that `E::A(x : τ) | E::B(x: τ)` should work if `x` on both sides are implicitly coercible to `τ` but I don't think we need to specify that interaction just yet.

Reply:
Amended the RFC with a clarification according to the above notes.

Reply:
`match` wouldn't be the coercion site here, *the individual patterns* would need to be. Currently, the `x` binding there *is the same variable* (initialized on two possible disjoint execution paths), and can't possibly have two types (but we can relax that in the future).